### PR TITLE
MAINT, BUG: Remove uses of PyString_FromString.

### PR DESCRIPTION
**Pull request by charris**
We no longer need to use the compatibility function after dropping support for
Python 2.7. In some cases unicode was the correct string type rather than the
bytes of the compatibility version and bugs in the array `__complex__` and
array `__array_interface__` methods have been fixed by changing that.

### CODE REVIEW

File: numpy/core/src/multiarray/methods.c [hunk 1]
```diff
@@ -2598,45 +2598,24 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
...
-        }
-        PyErr_SetObject(PyExc_TypeError, err);
-        Py_DECREF(err);
         return NULL;
     }
 
     if (PyArray_TYPE(self) == NPY_OBJECT) {
         /* let python try calling __complex__ on the object. */
         PyObject *args, *res;
-        Py_DECREF(dtype);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c:2633]: Why move this?

File: numpy/core/src/multiarray/common.c [hunk 2]
```diff
@@ -133,7 +133,7 @@ NPY_NO_EXPORT PyArray_Descr *
 _array_typedescr_fromstr(char const *c_str)
 {
     PyArray_Descr *descr = NULL;
-    PyObject *stringobj = PyString_FromString(c_str);
+    PyObject *stringobj = PyBytes_FromString(c_str);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.c:136]: This should be `PyUnicode_FromString`, otherwise it is a (likely  incorrect) behavior change.

File: numpy/core/src/multiarray/methods.c [hunk 3]
```diff
@@ -2598,45 +2598,24 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
 
     if (!PyArray_CanCastArrayTo(self, dtype, NPY_SAME_KIND_CASTING) &&
             !(PyArray_TYPE(self) == NPY_OBJECT)) {
-        PyObject *err, *msg_part;
+        PyObject *descr = (PyObject*)PyArray_DESCR(self);
+
+        PyErr_Format(PyExc_TypeError, "Unable to convert %R to complex", descr);
         Py_DECREF(dtype);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c:2604]: nit: move this up a line - decref things as soon as you're done with them.

File: numpy/core/src/multiarray/methods.c [hunk 4]
```diff
@@ -2598,45 +2598,24 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
...
-        }
-        msg_part = PyString_FromString(", to complex.");
-        if (msg_part == NULL) {
-            Py_DECREF(err);
-            return NULL;
-        }
-        PyString_ConcatAndDel(&err, msg_part);
-        if (err == NULL) {
-            return NULL;
-        }
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c:2624]: Thanks for cleaning this up - I expect most uses of `PyString_ConcatAndDel` can be replaced with a similar pattern now.

File: numpy/core/src/multiarray/methods.c (hunk 1 above)
Comment (charris) [reply numpy/core/src/multiarray/methods.c:2633]: Because we usually do clean just before returns. I find the code reads more logically this way as it wasn't immediately obvious why the reference was decremented where it was, you had to read through the code to see that it wasn't used and that there was a return.

File: numpy/core/src/multiarray/methods.c (hunk 3 above)
Comment (charris) [reply numpy/core/src/multiarray/methods.c:2604]: How do you know we are done with it until you read to the end?

File: numpy/core/src/multiarray/common.c (hunk 2 above)
Comment (charris) [reply numpy/core/src/multiarray/common.c:136]: Hmm... maybe. This doesn't change current behavior as this is the current definition of `PyString_FromString` for Python 3, but it may only be working by accident or not tested. @seberg Thoughts?

Comment (eric-wieser) [reply numpy/core/src/multiarray/common.c:136]: Hmm, I'm confused them. This patch is odd, because it replaces `PyString_FromString` with `PyUnicode_FromString` in some places, but `PyBytes_FromString` in others.

I didn't actually check which was which, but I assumed this one as the minority was the mistake.

File: numpy/core/src/multiarray/methods.c (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c:2624]: Oh, this fixes a real bug:
```python
In [127]: complex(np.array("now", np.datetime64))
TypeError: can't concat str to bytes
```

File: numpy/core/src/multiarray/methods.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c:2633]: Well as written, this now leaks `dtype` from line 2636 ;)

File: numpy/core/src/multiarray/methods.c (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c:2604]: My motivation is to make a statement of intent, not to write cautious code. If our goal was to be cautious, then we'd use `XDECREF` everywhere.

**Comment by eric-wieser** [comment]
Apologies @charris, I thought I was reviewing code from @jakobjakobson13! Identicons are apparently too similar at the end of the day.

File: numpy/core/src/multiarray/common.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/common.c:136]: Hmmm... This function has only two callers:

https://github.com/numpy/numpy/blob/7ec2e1bac72afcdc68cf8256879afbc4cb14a907/numpy/core/src/multiarray/ctors.c#L1779-L1781

https://github.com/numpy/numpy/blob/7ec2e1bac72afcdc68cf8256879afbc4cb14a907/numpy/core/src/multiarray/ctors.c#L1892-L1910

The first clearly doesn't care about string vs unicode (and the latter would be faster, because `PyArray_DescrConverter` starts by converting `str` to unicode).

The second seems a bit silly, as we take a unicode object, encode it to bytes, convert to char*, and then do the whole lot in reverse again.

Basically, this change is fine and I was wrong to think it a behavior change - but there's more cleanup that could be done here.

Comment (charris) [reply numpy/core/src/multiarray/common.c:136]: > This patch is odd, because it replaces PyString_FromString with PyUnicode_FromString

Yes. The other places looked like either would work and I thought unicode would be the better choice now that Python 2.7 support is gone.

File: numpy/core/src/multiarray/getset.c [hunk 5]
```diff
@@ -217,7 +217,7 @@ array_protocol_descr_get(PyArrayObject *self)
     if (dobj == NULL) {
         return NULL;
     }
-    PyTuple_SET_ITEM(dobj, 0, PyString_FromString(""));
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/getset.c:220]: This one was a bug too:
```python
>>> np.array(1).__array_interface__["descr"]
[('', '<i4')]
>>> np.array((1, 1), dtype=np.dtype(dict(names=['a', 'b'], offsets=[0, 0], formats=[int, int]))).__array_interface__['descr']
[(b'', '|V4')]  # different type!
```

File: numpy/f2py/cfuncs.py [hunk 6]
```diff
@@ -320,10 +320,10 @@
     'pyobj_from_complex_float1'] = '#define pyobj_from_complex_float1(v) (PyComplex_FromDoubles(v.r,v.i))'
 needs['pyobj_from_string1'] = ['string']
 cppmacros[
-    'pyobj_from_string1'] = '#define pyobj_from_string1(v) (PyString_FromString((char *)v))'
```
Comment (eric-wieser) [inline numpy/f2py/cfuncs.py:323]: This one worries me a bit, since the rest of this file uses `PyString_Check` a lot, which might no longer be satisfied.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c [hunk 7]
```diff
@@ -149,9 +149,9 @@ PyMODINIT_FUNC PyInit_test_array_from_pyobj_ext(void) {
   if (PyErr_Occurred())
     Py_FatalError("can't initialize module wrap (failed to import numpy)");
   d = PyModule_GetDict(m);
-  s = PyString_FromString("This module 'wrap' is auto-generated with f2py (version:2_1330).\nFunctions:\n"
-"  arr = call(type_num,dims,intent,obj)\n"
-".");
+  s = PyUnicode_FromString("This module 'wrap' is auto-generated with f2py (version:2_1330).\nFunctions:\n"
+                           "  arr = call(type_num,dims,intent,obj)\n"
+                           ".");
```
Comment (eric-wieser) [inline numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: Hmm, if this is auto-generated then maybe we should not be editing it?

File: numpy/core/src/multiarray/methods.c (hunk 1 above)
Comment (charris) [reply numpy/core/src/multiarray/methods.c:2633]: Fixed. The whole function can be rewritten but it doesn't look any prettier when I do.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c (hunk 7 above)
Comment (charris) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: It's in the repo, doesn't go away with `git clean -xdfq` and the content dates back to 2005. I think the string is intended for code generated by `wrapper.c`. However, in that case it is probably best to keep the message in bytes...

File: numpy/f2py/cfuncs.py (hunk 6 above)
Comment (charris) [reply numpy/f2py/cfuncs.py:323]: I was a also a bit uncertain about that. I think it is intended for returning strings to Python. Not sure how to check it without looking at some generated wrappers. @pearu Thoughts?

Comment (charris) [reply numpy/f2py/cfuncs.py:323]: I suppose it could cause problems if the user has worked around returned byte strings.

**Comment by eric-wieser** [comment]
It's tempting to split this into two phases:

* Perform dumb text substitution of all the compat macros in one PR
* Comb through the uses of `PyBytes_...` to check for things which don't make sense. I can't think of many situations where bytes are the correct choice in numpy.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c (hunk 7 above)
Comment (eric-wieser) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: > I think the string is intended for code generated by 

I think you're right, but the top of this file also says its autogenerated.

Comment (charris) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: Probably just left over from before the tests were redone. Note that it is only used in setting up `test_array_from_pyobj.py`, so I think it really doesn't matter.

**Comment by charris** [comment]
> It's tempting to split this into two phases:

I agree that we need to replace all the compat functions, but would rather do it in smaller pieces. Look how much work this small bit is already taking :) I think blind replacement just makes it harder to find the uses that may need fixing.

**Comment by seberg** [comment]
Right now, I am still happy to keep it here, since its not time pressing to do the replacement and the PR is not big.

The remaining issue is mainly the `f2py` change, right? It sounds like most of the other actual changes were fairly clear to resolve?

I suppose for some of the things that end up being fixes, including a test might be nice (also helps with judging whether anything should have release notes), but not sure.

**Review by Aryan05 — APPROVED** [review]

File: numpy/f2py/cfuncs.py (hunk 6 above)
Comment (charris) [reply numpy/f2py/cfuncs.py:323]: I've looked at this some more and think it the right thing to do here. It is a change, but I think likely harmless, and if it actually turns up in applications I think the fix would probably improve the code.

Comment (seberg) [reply numpy/f2py/cfuncs.py:323]: Thanks, should we add a small release note in that case (also helps me with understanding what the change is)?

**Comment by pearu** [comment]
I just noticed this PR and I'd like to review the f2py bits. However, it's late here but I can review tomorrow.

**Comment by charris** [comment]
@pearu Great. Other functions proposed for replacement in #17097 also appear in the f2py code so this may not be the last time review is needed.

File: numpy/f2py/cfuncs.py (hunk 6 above)
Comment (pearu) [reply numpy/f2py/cfuncs.py:323]: Scanning the files for `pyobj_from_` shows that the macro `pyobj_from_string1` is likely never used in wrapper generation, only `pyobj_from_string1size` and `pyobj_from_p_string1` are used.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c (hunk 7 above)
Comment (pearu) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: The `wrapmodule.c` is used in `numpy/f2py/tests/test_array_from_pyobj.py` and can be considered constant, that is, it can be manually modified.
For background: he file was generated with f2py as a C/API module template and later modified manually to expose various macro constants to Python for testing..

File: numpy/f2py/rules.py [hunk 8]
```diff
@@ -202,7 +202,7 @@
 \tif (PyErr_Occurred())
 \t\t{PyErr_SetString(PyExc_ImportError, \"can't initialize module #modulename# (failed to import numpy)\"); return m;}
 \td = PyModule_GetDict(m);
-\ts = PyString_FromString(\"$R""" + """evision: $\");
+\ts = PyUnicode_FromString(\"$R""" + """evision: $\");
```
Comment (pearu) [inline numpy/f2py/rules.py:205]: The `$Revision: $` is a relict from subversion days and the line can be replaced with
```
\ts = PyUnicode_FromString(\"#f2py_version#\");
```
While doing that, I suggest inserting the line
```
f2py_version = numpy_version
```
in the header of the rules.py file because the f2py version is de facto defined by numpy version.

**Review by pearu — APPROVED** [review]
LGTM from the f2py point of view.

While slightly OT, added some suggestions for updating f2py version handling.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c (hunk 7 above)
Comment (charris) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:154]: Sounds like that part of the header can just be removed. I'll leave this PR as is with a release note and make a separate PR with the version changes and other small cleanups suggested here.

**Comment by eric-wieser** [comment]
Think it's worth adding a test for the code snippets I give above that expose what this fixed?

File: numpy/f2py/cfuncs.py (hunk 6 above)
Comment (charris) [reply numpy/f2py/cfuncs.py:323]: Added a small release note.

**Comment by charris** [comment]
> Think it's worth adding a test for the code snippets I give above

Been thinking about that. We could add those to the regressions checks, but in the long run it would be useful to figure out how to generate a set of comprehensive tests for both errors and descriptors. @seberg may have some ideas about the latter.

**Review by eric-wieser — APPROVED** [review]
Thanks for the tests. I think with @pearu's review of the f2py stuff, this is now good to go.

File: numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c (hunk 7 above)
Comment (mattip) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:149]: Would you like to wait for this to be merged until the follow-on PR is made?

Comment (eric-wieser) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:149]: I don't understand your question @mattip, the header is changed already.

Comment (charris) [reply numpy/f2py/tests/src/array_from_pyobj/wrapmodule.c:149]: @mattip No :) I'm waiting for this to be merged. The follow-on PR just changes the version handling.

**Comment by mattip** [comment]
I was asking generally about this PR: there are follow-on cleanups:
- in `numpy/numpy/core/src/multiarray/ctors.c` where we call `_array_typedescr_fromstr` after converting a str to bytes then back to str
- changes in line 205 to `numpy/f2py/rules.py` to use `f2py_version`and delete line 53

I guess we should merge this now and then submit a follow-on PR. In it goes.

**Comment by mattip** [comment]
Thanks @charris
